dotnet-desktop-guide/net/wpf/documents/snippets/how-to-display-print-dialog/csharp/MainWindow.xaml.cs

```diff
@@ -0,0 +1,227 @@
+
+                // Get a fixed document sequence for the selected file
+                FixedDocumentSequence fixedDocSeq = xpsDocument.GetFixedDocumentSequence();
+
+                // Create a paginator for all pages in the selected file
+                DocumentPaginator docPaginator = fixedDocSeq.DocumentPaginator;
+
+                // Print to a new file
+                printDialog.PrintDocument(docPaginator, $"Printing {Path.GetFileName(xpsFilePath)}");
+            }
+            catch (Exception e)
+            {
+                MessageBox.Show(e.Message);
+            }
+        }
+        // </SampleCode>
+
+        /// <summary>
+        /// Print an XPS document to OXPS format. Optionally, print a specific range
+        /// of pages using a print dialog window, or all pages without the dialog.
+        /// </summary>
+        /// <param name="xpsFilePath">Path to source XPS file</param>
+        /// <param name="useDialog">Whether to launch a Print Dialog</param>
+        private static void PrintFileWithSpecificPages(string xpsFilePath, bool useDialog)
+        {
+            // Create the print dialog object and set options
+            PrintDialog printDialog = new()
+            {
+                UserPageRangeEnabled = useDialog
+            };
+
+            // Display the dialog. This returns true if the user presses the Print button
+            if (useDialog)
+            {
+                bool? isPrinted = printDialog.ShowDialog();
+                if (isPrinted != true)
+                    return;
+            }
+
+            // Print the document to an XPS or OXPS file
+            try
+            {
+                // Open the selected file
+                XpsDocument xpsDocument = new(xpsFilePath, FileAccess.Read);
+
+                // Get a fixed document sequence for the selected file
+                FixedDocumentSequence fixedDocSeq = xpsDocument.GetFixedDocumentSequence();
+
+                // Create a paginator for all pages in the selected file
+                DocumentPaginator docPaginator = fixedDocSeq.DocumentPaginator;
+
+                // Check whether a page range was specified in the print dialog
+                if (printDialog.PageRangeSelection == PageRangeSelection.UserPages)
+                {
+                    // Create a document paginator for the specified range of pages
+                    docPaginator = new DocPaginator(fixedDocSeq.DocumentPaginator, printDialog.PageRange);
+                }
+
+                // Print to a new file
+                printDialog.PrintDocument(docPaginator, $"Printing {Path.GetFileName(xpsFilePath)}");
+            }
+            catch (Exception e2)
+            {
+                MessageBox.Show(e2.Message);
+            }
+        }
+
+        /// <summary>
+        /// Extend the abstract DocumentPaginator to support printing specific page ranges
+        /// </summary>
+        public class DocPaginator : DocumentPaginator
```

Comment:
If you feel up to it, you could alter the `## Example` heading to be something like `## Print the whole document` and then add a new header `## Print a page range` and then add some new content to describe how to do this. I think it would be very helpful to the customers and, honestly, a key scenario that's been left out of the original documentation.

Reply:
I'll add a new header as you suggest. I agree that providing a working solution to page range printing will be helpful to customers. It's too bad one of the XPS developers isn't available to review and provide an official approach to that exception workaround code.

Reply:
Done.

Reply:
I think that this was left out of WPF because they expect you to do it all by hand. If there was a straightforward way, we would have discovered it by now, but looking at the WPF samples repo, it's not simple, this isn't a workaround, it's a way of implementing it lol :)

Reply:
I agree the extended class `DocPaginator` is unavoidable since the class it extends the abstract (`DocumentPaginator`) class. 

The workaround that I'm not liking is just the code section starting with:

> // Workaround for "FixedPageInPage" exception.

I've used Microsoft code snippets quite a bit for my own projects over the years, so I've a pretty good sense when code doesn't look right, and this is one of those rare cases that would make me wonder about the knowledge level of the author. However, I agree it's better than providing nothing at all. :)

Reply:
Specifically, it's the use of reflection that is problematic. For example, it's harder to optimize code, i.e. strip-out compiler injected metadata, when reflection requires that metadata.